feat(log): Add finer grained logging

rasa opened the pull request:
This PR allows one to set the STTRACE environment variable as:

  `facility[:level][,facility2[:level2]] ...`

For example:

  `all:warn,fs:info,model`

will log everything at the WARN level (so no INFO lines in the logs), logs the fs facility at the INFO level (no DEBUG lines), and logs the model facility at the DEBUG level.

Abbreviations are allowed, so

  `all:w,fs:i,model`

is the same as above.

The PR is backwards compatible with the existing logic.

### Purpose

I often run my Syncthing logs thru grep to filter out the INFO lines, when I'm debugging. This allows me to simplify (or do away with) my grep statements.

### Testing

I added unit tests to the new functionality.

### Documentation

If this PR is accepted, I'm draft a PR to update the docs.

## Authorship

My name's already in our AUTHORS file.

rasa commented:
All tests passed locally on both Ubuntu and Windows 11, so I'm not sure why they are failing on GitHub. I will research. Until then, I'll mark it as draft.

acolomb commented:
Please don't forget to adjust the documentation to cover the new syntax.

imsodin commented on lib/logger/logger.go:
```diff
@@ -34,8 +33,24 @@ const (
 const (
 	DefaultFlags = log.Ltime | log.Ldate
 	DebugFlags   = log.Ltime | log.Ldate | log.Lmicroseconds | log.Lshortfile
+	LevelError   = NumLevels // Not really a log level, as it disables all logging.
```
I'd prefer to have something dedicated to disabling logging, not using the fact that we don't do error logging as a standin. I was initially confused, as I did expect error logging to exist. And obviously not your PR, but I think it should exist: Imo it would make sense to use an error level for fatal/grave errors distinct from unexpected, but recoverable/handleable errors - currently that's all warning level. And using error level for disabling in this PR likely makes that harder, when/if we want to use error level logging in the future. I haven't checked the diff properly, but from a brief look it doesn't seem like you need to change much, if any logic, it might just be a comment and possibly naming issue to avoid equating error logging to disabling logging entirely.

rasa replied:
@imsodin I agree 100%. I will change it to `LevelDisabled`. I didn't want to reuse `NumLevels` as that doesn't describe its function in this context.

With this PR's changes, it would be trivial to add an actual `LevelError` or `LevelFatal`. I didn't want to expand the scope of this PR, but happy to do so.

imsodin replied:
Very much agree with keeping the scope, not mixing in more things.

rasa commented:
@imsodin Great feedback. I implemented your suggestion, along with a few other tweaks:
1. Renamed `setFlags`, to `setLoggerFlags` to quiet DeepSource's [complaint](https://app.deepsource.com/gh/syncthing/syncthing/run/990ef3a8-6a67-4588-903c-9f6022b6fb1b/go/RVV-B0001?listcategory&listsort&listq&listindex=0).
2. Added `LevelDefault` alias to `LevelInfo` to clearly indicate `LevelInfo` is the default. This could be private, if desired.
3. Moved delimiter characters to a constant to share with test suite.
4. Fixed typos in comments.

calmh commented:
IMHO we should not do this... I think the concept is valid, but building additional machinery on top of our existing logging framework is beating the dead horse. I have on my mental todo list at some point to switch to the more modern slog package/approach, at which point this could be one part of that change...
